### PR TITLE
Custom BlocksBlock attributes

### PR DESCRIPTION
**Pull request by johnnyomair**
Introduces the ability to add additional attributes for child blocks of a BlocksBlock. This can be useful to add custom features, for instance visibility rules, to a BlocksBlock.

# Demo

https://user-images.githubusercontent.com/48853629/187610142-659e560f-25a5-441f-8b34-e094f8c004a8.mov

# Custom Indicator

Indicator

### CODE REVIEW

**Comment by johnnyomair** [comment]
> isn't it limitating to have a 1:1 relation from additionalField values and ContextMenu entires?

Technically, it's not a 1:1 relation, as the `ContextMenuItem` component doesn't expect a single MenuItem. We might change the naming to avoid this confusion.

**Comment by johnnyomair** [comment]
@nsams I like the necessary changes in the API, but I'm not so fond of the Admin implementation. Having `Indicator` and `ContextMenuItem` components is rather specific. Do you have any ideas how this could be more flexible? Should we be able to render a custom BlockRow instead?

**Comment by johnnyomair** [comment]
@nsams I've tried to loosen the relation between additional fields and UI components bei moving the components up to the BlocksBlock level: 03da69d30a668efeed31e03ffd4839c21adeb542. What do you think?

File: demo/admin/src/pages/PageContentBlock.tsx [hunk 1]
```diff
@@ -20,4 +40,73 @@ export const PageContentBlock = createBlocksBlock({
...
+                    onClose={() => {
+                        setDialogOpen(false);
+                        onMenuClose();
+                    }}
+                >
+                    <DialogTitle>
+                        <FormattedMessage id="cometDemo.pageContentBlock.userGroup.dialogTitle" defaultMessage="Visibility rules" />
+                    </DialogTitle>
+                    <DialogContent>
+                        <Select value={internalValue} onChange={(event) => setInternalValue(event.target.value)} fullWidth>
```
Comment (nsams) [inline demo/admin/src/pages/PageContentBlock.tsx:74]: why not use final-form here?

File: demo/admin/src/pages/PageContentBlock.tsx [hunk 2]
```diff
@@ -20,4 +40,73 @@ export const PageContentBlock = createBlocksBlock({
         linkList: LinkListBlock,
         fullWidthImage: FullWidthImageBlock,
     },
+    additionalFields: {
+        userGroup: {
+            defaultValue: "All",
+        },
+    },
+    AdditionalContextMenuItems: ({ block, onChange, onMenuClose }) => {
```
Comment (nsams) [inline demo/admin/src/pages/PageContentBlock.tsx:48]: what is `block`, the block state?

Comment (johnnyomair) [reply demo/admin/src/pages/PageContentBlock.tsx:48]: It's a single child of the blocks block. Internally we call this `blocksBlockItem`. I may change it here to `item`, but that may be confusing with `AdditionalContextMenuItems`. Maybe `blocksBlockItem`?

File: demo/admin/src/pages/PageContentBlock.tsx (hunk 1 above)
Comment (johnnyomair) [reply demo/admin/src/pages/PageContentBlock.tsx:74]: I've actually not thought about using it, as it's only a single select. But you're right, this is better than keeping track of an internal state

File: demo/admin/src/pages/PageContentBlock.tsx (hunk 2 above)
Comment (nsams) [reply demo/admin/src/pages/PageContentBlock.tsx:48]: > Maybe blocksBlockItem?

:+1:

File: demo/admin/src/pages/PageContentBlock.tsx (hunk 1 above)
Comment (johnnyomair) [reply demo/admin/src/pages/PageContentBlock.tsx]: Now using Final Form: d753ae7a17106d5ddabf9eb00d7f002015139dda

**Comment by johnnyomair** [comment]
TODO (maybe in a follow-up PR): Type inference for additional fields

**Review by chernylu — APPROVED** [review]

**Review by raphaelblum — APPROVED** [review]

**Review by nsams — APPROVED** [review]

**Pull request merged**